File: apps/swift-demo/Sources/SwiftDemoApp/Store.swift
import Foundation
import Antifragile

// ---------------------------------------------------------------------------
// Task — plain struct, no store dependency. Tested with plain init.
// ---------------------------------------------------------------------------

struct Task: Identifiable, Equatable {
    let id: String
    let title: String
    let status: String   // "active" | "archived"
    let createdAt: TimeInterval

    var createdAtDisplay: String {
        Date(timeIntervalSince1970: createdAt)
            .formatted(.dateTime.month(.abbreviated).day().year())
    }

    var statusLabel: String {
        status == "active" ? "In Progress" : "Archived"
    }

    static func from(_ doc: [String: Any]) -> Task? {
        guard
            let id        = doc["id"]        as? String,
            let title     = doc["title"]     as? String,
            let status    = doc["status"]    as? String,
            let createdAt = doc["createdAt"] as? TimeInterval
        else { return nil }
        return Task(id: id, title: title, status: status, createdAt: createdAt)
    }
}

typealias TaskId = String

// ---------------------------------------------------------------------------
// Mutates — named write descriptors, standalone
// ---------------------------------------------------------------------------

let addTask = createMutate(action: "AddTask") { payload in
    guard let id    = payload["id"]    as? String,
          let title = payload["title"] as? String else { return [] }
    return [Write(path: "tasks", id: id, fields: [
        "title":     title,
        "status":    "active",
        "createdAt": Date().timeIntervalSince1970,
    ])]
}

let archiveTask = createMutate(action: "ArchiveTask") { payload in
    guard let id = payload["id"] as? String else { return [] }
    return [Write(path: "tasks", id: id, fields: ["status": "archived"])]
}

// ---------------------------------------------------------------------------
// Store — seed data + mutates at the bottom
// ---------------------------------------------------------------------------

let store = Store.createStore {
    BackingStoreConfig(
        name: "default",
        adapter: MemoryAdapter(initial: [
            "tasks": [
                "task-1": [
                    "id": "task-1",
                    "title": "Deploy to production",
                    "status": "active",
                    "createdAt": Date().timeIntervalSince1970 - 86_400,
                ],
                "task-2": [
                    "id": "task-2",
                    "title": "Write release notes",
                    "status": "active",
                    "createdAt": Date().timeIntervalSince1970 - 3_600,
                ],
                "task-3": [
                    "id": "task-3",
                    "title": "Update dependencies",
                    "status": "active",
                    "createdAt": Date().timeIntervalSince1970,
                ],
            ],
        ]),
        mutates: [addTask, archiveTask]
    )
}

File: apps/ts-demo/src/store.ts
import { createStore, createWireView } from '@fiskal/antifragile'
import { MemoryAdapter } from '@fiskal/antifragile/adapters/memory'
import type { Model } from '@fiskal/antifragile'

// ---------------------------------------------------------------------------
// Model — JSON schema + closure-based compute
// ---------------------------------------------------------------------------

const taskCompute: Model['compute'] = {
  statusLabel:      (doc) => doc['status'] === 'active' ? 'In Progress' : 'Archived',
  createdAtDisplay: (doc) => new Date(doc['createdAt'] as number).toLocaleDateString(undefined, {
    month: 'short', day: 'numeric', year: 'numeric',
  }),
}

// ---------------------------------------------------------------------------
// Store — seed data, model, and mutates all inline
// ---------------------------------------------------------------------------

export const store = createStore(
  MemoryAdapter({
    tasks: [
      { id: 'task-1', title: 'Deploy to production', status: 'active', createdAt: Date.now() - 86_400_000 },
      { id: 'task-2', title: 'Write release notes',  status: 'active', createdAt: Date.now() - 3_600_000  },
      { id: 'task-3', title: 'Update dependencies',  status: 'active', createdAt: Date.now()              },
    ],
  }),
  {
    models: {
      tasks: {
        schema: {
          type: 'object',
          properties: {
            id:        { type: 'string' },
            title:     { type: 'string', minLength: 1 },
            status:    { type: 'string', enum: ['active', 'archived'] },
            createdAt: { type: 'number' },
          },
          required: ['id', 'title', 'status', 'createdAt'],
        },
        compute: taskCompute,
      },
    },
    mutates: {
      addTask: {
        write: (payload) => {
          const { id, title } = payload as { id: string; title: string }
          return {
            path: 'tasks',
            id,
            fields: { title, status: 'active', createdAt: Date.now() },
          }
        },
      },
      archiveTask: {
        write: (payload) => {
          const { id } = payload as { id: string }
          return { path: 'tasks', id, fields: { status: 'archived' } }
        },
      },
    },
  },
)

export const wireView = createWireView(store)
